dg.go

```diff
@@ -1,315 +1,464 @@
+	for nodeID, n := range d.nodes {
+		result[nodeID] = n
+	}
+	return result
 }
 
 func (d *directedGraph[NodeType]) ListNodesWithoutInboundConnections() map[string]Node[NodeType] {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        if len(d.connectionsToNode[nodeID]) == 0 {
-            result[nodeID] = n
-        }
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		connections := d.connectionsToNode[nodeID]
+		if len(connections) == 0 {
+			result[nodeID] = n
+		}
+	}
+	return result
+}
+
+func (d *directedGraph[NodeType]) connectNodes(fromID, toID string, dependencyType DependencyType) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	// Make sure both nodes exist and are not deleted.
+	fromNode, ok := d.nodes[fromID]
+	if !ok {
+		return &ErrNodeNotFound{fromID}
+	} else if fromNode.deleted {
+		return &ErrNodeDeleted{fromID}
+	}
+	toNode, ok := d.nodes[toID]
+	if !ok {
+		return &ErrNodeNotFound{
+			toID,
+		}
+	} else if toNode.deleted {
+		return &ErrNodeDeleted{toID}
+	}
+	// Validate that it's a valid, non-duplicate, connection.
+	if fromID == toID {
+		return &ErrCannotConnectToSelf{
+			fromID,
+		}
+	}
+	if _, ok := d.connectionsFromNode[fromID][toID]; ok {
+		return &ErrConnectionAlreadyExists{
+			fromID,
+			toID,
+		}
+	}
+	// Update the mappings.
+	d.connectionsFromNode[fromID][toID] = struct{}{}
+	d.connectionsToNode[toID][fromID] = struct{}{}
+	// Update the dependencies
+	toNode.outstandingDependencies[fromID] = dependencyType
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PushStartingNodes() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	for nodeID, n := range d.nodes {
+		if len(n.outstandingDependencies) == 0 {
+			d.readyForProcessing[nodeID] = n
+		}
+	}
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PopReadyNodes() []*node[NodeType] {
+	d.lock.Lock()
+	// Save the map to a local variable to minimize time locked.
+	readyMap := d.readyForProcessing
+	d.readyForProcessing = make(map[string]*node[NodeType])
+	d.lock.Unlock()
+	// Move the map nodes to a slice.
+	result := make([]*node[NodeType], len(readyMap))
+	i := 0
+	for _, node := range readyMap {
+		result[i] = node
+		i += 1
+	}
+	return result
 }
 
 type node[NodeType any] struct {
-    deleted bool
-    id      string
-    item    NodeType
-    dg      *directedGraph[NodeType]
+	deleted                 bool
+	id                      string
+	item                    NodeType
+	ready                   bool
+	status                  ResolutionStatus
+	outstandingDependencies map[string]DependencyType
+	dg                      *directedGraph[NodeType]
+}
+
+func (n *node[NodeType]) ID() string {
+	return n.id
+}
+
+func (n *node[NodeType]) Item() NodeType {
+	return n.item
+}
+
+func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
+	return n.status
+}
```

Comment:
Added locks to the appropriate getters.

Reply:
But the bigger question here is *why* would it be locked?
1. Because Go's compilation of `return n.status` isn't atomic, and needs to be protected (e.g. against deletion of the node);
2. Because there are possible dangerous states of `n.status` which aren't a valid `ResolutionStatus` (which seems unlikely);
3. Because you're protecting the relationship between multiple predicates including `n.status` ... in which case locking *here* is of no use.

Reply:
In the absence of a lock, we're subject to software and hardware caching and visibility issues, possibly resulting in the return of stale values; however, while adding the locks prevents/addresses those issues, the fact that we release the lock before returning the value means that by the time we get to the caller, the value may again be stale (due to a change by some other thread).  So, we're potentially damned either way.

The only way to win is to figure out how this function is supposed to be used and to ensure that the appropriate precautions are used.  I believe that that means that `status` has to be read and the appropriate decision has to be reached without releasing the lock in between...which means that adding locking to this function will likely get in the way.  The best approach would likely be to remove this function altogether and, instead, move whatever function wants to call it into a `node` or DAG receiver where it can lock the lock and perform its entire purpose.

Reply:
> software and hardware caching and visibility issues, possibly resulting in the return of stale values

Exactly -- as you point out, unless we're relating to another predicate under a common lock, we have no "liveness" guarantee, so none of that really matters. (*Maybe*, depending on a lot of factors we don't control and can't directly see, the lock might limit the staleness in some remotely useful way, but that's not much of a guarantee for the overhead.)

> I believe that that means that status has to be read and the appropriate decision has to be reached without releasing the lock in between

Yeah, that's what I expect, which is my (3). In this case, the isolated accessor, locked or not, is meaningless.